01. Importing libraries

In [59]:
# Import the pandas, NumPy, and os libraries
import pandas as pd
import numpy as np
import os

02. Importing Data

In [60]:
# Turning the project folder path into a string
path = r'/Users/fatemehshahvirdi/Work-Related/Data Analysis/Data Immersion/Achievement 4/Instacart Basket Analysis'

In [61]:
path

'/Users/fatemehshahvirdi/Work-Related/Data Analysis/Data Immersion/Achievement 4/Instacart Basket Analysis'

In [62]:
#importing dataset orders
df_ords = pd.read_csv( os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [63]:
#importing dataset products
df_prods = pd.read_csv( os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [64]:
#importing dataset departments
df_dep = pd.read_csv( os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

03. Data wrangling

In [115]:
df_prods.shape

(49693, 5)

In [66]:
df_ords.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [67]:
df_ords.drop(columns=['eval_set'], inplace=True)

In [68]:
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day',
       'days_since_prior_order'],
      dtype='object')

In [69]:
# Count the occurrences of each value in the 'days_since_prior_order' column,
# including missing values, and display the result
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [70]:
#rename the 'order_dow' to 'orders_day_od_week', and apply the changes directly to the data frame (inplace)
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'},inplace = True )

In [71]:
#show 5 raws of the orders data frame
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [72]:
#show the 20 first raws of the column'orders_of_the_week' from the orders data frame
print(df_ords['orders_day_of_week'].head(20))

0     2
1     3
2     3
3     4
4     4
5     2
6     1
7     1
8     1
9     4
10    4
11    2
12    5
13    1
14    2
15    3
16    2
17    2
18    1
19    2
Name: orders_day_of_week, dtype: int64


In [73]:
#convert the data type of the 'order_id' to a string 
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [74]:
df_ords['order_id'].dtype

dtype('O')

In [75]:
## Display the first few rows of the DataFrame df_dep.
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [76]:
#transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [77]:
# Transpose the DataFrame df_dep and assign it to df_dep_t.
df_dep_t = df_dep.T

In [78]:
#display 'df_dep.t'
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [79]:
## Reset the index of the transposed DataFrame df_dep_t.
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [80]:
# Set the first row of the transposed DataFrame df_dep_t as the new header
new_header = df_dep_t.iloc[0]

In [81]:
# Display the new header of the transposed DataFrame df_dep_t
print(new_header)

0    department
Name: department_id, dtype: object


In [82]:
# Create a new DataFrame 'df_dep_t_new' containing the data from 'df_dep_t', after the first row.
df_dep_t_new = df_dep_t[1:]

In [83]:
# Display the first raws ot the 'df_dep_t_new'
df_dep_t_new.head()

0
1   frozen
2    other
3   bakery
4  produce
5  alcohol

In [84]:
# Set the column names of 'df_dep_t_new'to the values in the new_header.
df_dep_t_new.columns = new_header

In [85]:
# Display the df_dep_t_new
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

04. Data Dictionary

In [86]:
# Turn your 'df_dep_t_new' data frame into a dictionary
data_dict = df_dep_t_new.to_dict('index')

In [87]:
# Display the 'data_dict'
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

05. Subsetting

In [88]:
# Display the first few rows of the df_prods DataFrame.
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [89]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [90]:
#create a subset, by filtering the DataFrame df_prods to the raws that only include the department_id column that equals 19
# indicating snacks.
df_snacks = df_prods[df_prods['department_id']==19]

In [91]:
# Display the first few rows of the df_snacks DataFrame
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [92]:
# Other ways of creating a subset
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [93]:
## Other ways of creating a subset
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

Task Procedure.   
   6.1.  Find another identifier variable in the df_ords data frame that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [94]:
# Checking the data in orders.csv
df_ords.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   
5  3367565        1             6                   2                  7   
6   550135        1             7                   1                  9   
7  3108588        1             8                   1                 14   
8  2295261        1             9                   1                 16   
9  2550362        1            10                   4                  8   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0

In [95]:
# user_id should be changed to the string, since it doesn't have numeric value
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [96]:
# Checking the datatype of the user_id in orders.csv
df_ords['user_id'].dtype

dtype('O')

6.2. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [97]:
# checking the orders data frame
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [98]:
df_ords.rename(columns = {'order_number' : 'order_sequence_number'}, inplace = False)

order_id user_id  order_sequence_number  orders_day_of_week  \
0        2539329       1                      1                   2   
1        2398795       1                      2                   3   
2         473747       1                      3                   3   
3        2254736       1                      4                   4   
4         431534       1                      5                   4   
...          ...     ...                    ...                 ...   
3421078  2266710  206209                     10                   5   
3421079  1854736  206209                     11                   4   
3421080   626363  206209                     12                   1   
3421081  2977660  206209                     13                   1   
3421082   272231  206209                     14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

6.3. Your client wants to know what the busiest hour is for placing orders. 
Find the frequency of the corresponding variable and share your findings.

In [99]:
df_ords['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

The busiest hour is 10.

6.4. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [100]:
# Checking the data frame
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [101]:
print(data_dict.get('4'))

{'department': 'produce'}


6.5.The sales team in your client’s organization wants to know more about breakfast item sales. 
Create a subset containing only the required information.

In [102]:
# Checking the data dictionary to know which department is related to the breakfast item
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

department 14 is the breakfast item

In [103]:
# 1st Step fpr creating a subset
# Gathering data from 14 department_id
df_breakfast = df_prods [df_prods['department_id'] == 14 ]

In [104]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

6.6. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire data frame that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [105]:
# The first step is to find the numbers of 'department_id' for the ' alcohol, deli, beverages, and meat/seafood '
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

The related 'department_id', are: '5' for 'alcohol', '7' for' beverages', '12' for 'meat seafood', and '20' for 'deli'

In [106]:
# Creating a subset of related department_id
df_dinner_party = df_prods[ df_prods ['department_id'].isin([5,7,12,20])]

In [107]:
df_dinner_party.head(10)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5

6.7. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [108]:
df_dinner_party.shape

(7650, 5)

The last data frame that I created (dinner_party) has 7650 raws.

6.8. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [109]:
# The right data frame is the df_ords, and I need to create a new data frame since the user_id 1 is repeated multiple times in the data frame
df_customer_1 = df_ords[df_ords['user_id'] == 1]

In [110]:
# I want to procide them with the first and last data about customer with the user_id 1
df_customer_1.head()

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [111]:
df_customer_1.tail()

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

6.9. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [112]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

6.10. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments' data in the correct format.

In [113]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))